src/fs/metadata_log_bootstrap.hh

```diff
@@ -0,0 +1,488 @@
+#include "fs/inode_info.hh"
+#include "fs/metadata_disk_entries.hh"
+#include "fs/units.hh"
+#include "fs/metadata_log.hh"
+#include "seastar/core/do_with.hh"
+#include "seastar/core/future-util.hh"
+#include "seastar/core/future.hh"
+#include "seastar/core/sstring.hh"
+#include "seastar/core/temporary_buffer.hh"
+
+#include <boost/crc.hpp>
+#include <cstddef>
+#include <cstring>
+#include <unordered_set>
+#include <variant>
+
+namespace seastar::fs {
+
+class data_reader {
+    const uint8_t* _data;
+    size_t _size;
+    size_t _pos;
+
+public:
+    data_reader() : _data(nullptr), _size(0), _pos(0) {}
+
+    data_reader(const uint8_t* data, size_t size) : _data(data), _size(size), _pos(0) {}
+
+    size_t curr_pos() const noexcept { return _pos; }
+
+    size_t bytes_left() const noexcept { return _size - _pos; }
+
+    // Returns whether the reading was successful
+    bool read(void* destination, size_t size) {
+        if (_pos + size > _size) {
+            return false;
+        }
+
+        std::memcpy(destination, _data + _pos, size);
+        _pos += size;
+        return true;
+    }
+
+    // Returns whether the reading was successful
+    template<class T>
+    bool read_entry(T& entry) noexcept {
+        return read(&entry, sizeof(entry));
+    }
+
+    // Returns whether the reading was successful
+    bool read_sstring(sstring& str, size_t size) {
+        str.resize(size);
+        return read(str.data(), size);
+    }
+
+    std::optional<temporary_buffer<uint8_t>> read_tmp_buff(size_t size) {
+        if (_pos + size > _size) {
+            return std::nullopt;
+        }
+
+        _pos += size;
+        return temporary_buffer<uint8_t>(_data + _pos - size, size);
+    }
+
+    // Returns whether the processing was successful
+    bool process_crc_without_reading(boost::crc_32_type crc, size_t size) {
+        if (_pos + size > _size) {
+            return false;
+        }
+
+        crc.process_bytes(_data + _pos, size);
+        return true;
+    }
+
+    std::optional<data_reader> extract(size_t size) {
+        if (_pos + size > _size) {
+            return std::nullopt;
+        }
+
+        _pos += size;
+        return data_reader(_data + _pos - size, size);
+    }
+};
+
+class metadata_log_bootstrap {
+    metadata_log& _metadata_log;
+    cluster_range _available_clusters;
+    std::unordered_set<cluster_id_t> _taken_clusters;
+    std::optional<cluster_id_t> _next_cluster;
+    temporary_buffer<uint8_t> _curr_cluster_data;
+    data_reader _curr_cluster;
+    data_reader _curr_checkpoint;
+
+    metadata_log_bootstrap(metadata_log& metadata_log, cluster_range available_clusters)
+    : _metadata_log(metadata_log)
+    , _available_clusters(available_clusters)
+    , _curr_cluster_data(decltype(_curr_cluster_data)::aligned(metadata_log._alignment, metadata_log._cluster_size))
+    {}
+
+    future<> bootstrap(cluster_id_t first_metadata_cluster_id, fs_shard_id_t fs_shards_pool_size, fs_shard_id_t fs_shard_id) {
```

Comment:
Note: the code below looks fine, but consider simply spawning a seastar thread with `seastar::async`. Bootstrapping is a rare operation and `seastar::async` makes the code so much more readable, plus cancels the need of `do_with`s and other liveness-related tricks.

Reply:
Does seastar::async have any performance impact beyond small memory footprint? If so, I think that the code is in pretty good shape.

Reply:
`seastar::async` needs to allocate a rather large stack - currently 128KiB, but it's going to be customizable soon. On the other hand, the stack is allocated once, so it might sometimes be beneficial, compared to a complicated chain of futures, each requesting a small allocation.

Once we have coroutines, that would be the third viable alternative - which will have one small allocation for all needed variables.